velox/functions/sparksql/String.cpp

```diff
@@ -103,6 +103,273 @@ class Length : public exec::VectorFunction {
+        continue;
+      }
+      if (args[i] && args[i]->as<ConstantVector<StringView>>() &&
+          !args[i]->as<ConstantVector<StringView>>()->isNullAt(0)) {
+        std::ostringstream out;
+        out << args[i]->as<ConstantVector<StringView>>()->valueAt(0);
+        column_index_t j = i + 1;
+        // Concat constant string args in advance.
+        for (; j < args.size(); ++j) {
+          if (!args[j] || args[j]->typeKind() == TypeKind::ARRAY ||
+              !args[j]->as<ConstantVector<StringView>>() ||
+              args[j]->as<ConstantVector<StringView>>()->isNullAt(0)) {
+            break;
+          }
+          out << separator_.value()
+              << args[j]->as<ConstantVector<StringView>>()->valueAt(0);
+        }
+        constantStrings.emplace_back(out.str());
+        i = j - 1;
+      } else {
+        constantStrings.push_back("");
+      }
+    }
+
+    // Number of string columns after combined consecutive constant ones.
+    auto numStringCols = constantStrings.size();
+    for (auto i = 0; i < numStringCols; ++i) {
+      if (constantStrings[i].empty()) {
+        auto index = argMapping[i];
+        decodedStringArgs.emplace_back(context, *args[index], rows);
+      }
+    }
+  }
+
+  // ConcatWs implementation. It concatenates the arguments with the separator.
+  // Mixed using of VARCHAR & ARRAY<VARCHAR> is considered. If separator is
+  // constant, concatenate consecutive constant string args in advance. Then,
+  // concatenate the intemediate result with neighboring array args or
+  // non-constant string args.
+  void doApply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      exec::EvalCtx& context,
+      FlatVector<StringView>& flatResult) const {
+    std::vector<column_index_t> argMapping;
+    std::vector<std::string> constantStrings;
+    auto numArgs = args.size();
+    argMapping.reserve(numArgs - 1);
+    // Save intermediate result for consecutive constant string args.
+    // They will be concatenated in advance.
+    constantStrings.reserve(numArgs - 1);
+    std::vector<exec::LocalDecodedVector> decodedArrays;
+    decodedArrays.reserve(numArgs - 1);
+    // For column string arg decoding.
+    std::vector<exec::LocalDecodedVector> decodedStringArgs;
+    decodedStringArgs.reserve(numArgs);
+
+    initVectors(
+        rows,
+        args,
+        context,
+        decodedArrays,
+        argMapping,
+        constantStrings,
+        decodedStringArgs);
+    exec::LocalDecodedVector decodedSeparator(context);
+    if (!separator_.has_value()) {
+      decodedSeparator = exec::LocalDecodedVector(context, *args[0], rows);
+    }
+
+    auto totalResultBytes = calculateTotalResultBytes(
+        rows,
+        context,
+        decodedArrays,
+        constantStrings,
+        decodedStringArgs,
+        decodedSeparator);
+
+    // Allocate a string buffer.
+    auto rawBuffer = flatResult.getRawStringBufferWithSpace(totalResultBytes);
```

Comment:
Is `totalResultBytes` the exact size? Maybe use `getRawStringBufferWithSpace(totalResultBytes, true)` if it is.

Reply:
Fixed. Thanks!